File: app/src/main/java/com/usharik/app/AppState.java
package com.usharik.app;

import androidx.lifecycle.MutableLiveData;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by macbook on 07/03/2018.
 * Application state management using LiveData for reactive updates.
 * This class maintains the global state of the quiz application.
 */
public class AppState {
    private final MutableLiveData<Map<String, Integer>> wordsWithErrorsLiveData = new MutableLiveData<>(new HashMap<>());
    private final MutableLiveData<String> genderFilterStrLiveData = new MutableLiveData<>(Gender.ALL);
    private final MutableLiveData<Integer> genderFilterIdLiveData = new MutableLiveData<>(-1);
    private final MutableLiveData<Boolean> switchOffAnimationLiveData = new MutableLiveData<>(false);

    public Map<String, Integer> getWordsWithErrors() {
        Map<String, Integer> value = wordsWithErrorsLiveData.getValue();
        return value != null ? value : new HashMap<>();
    }

    public String getGenderFilterStr() {
        String value = genderFilterStrLiveData.getValue();
        return value != null ? value : Gender.ALL;
    }

    public int getGenderFilterId() {
        Integer value = genderFilterIdLiveData.getValue();
        return value != null ? value : -1;
    }

    public boolean getSwitchOffAnimation() {
        Boolean value = switchOffAnimationLiveData.getValue();
        return value != null ? value : false;
    }

    public void setWordsWithErrors(Map<String, Integer> wordsWithErrors) {
        this.wordsWithErrorsLiveData.setValue(wordsWithErrors);
    }

    public void setSwitchOffAnimation(boolean switchOffAnimation) {
        this.switchOffAnimationLiveData.setValue(switchOffAnimation);
    }

    public void setGenderFilterStr(String genderFilterStr) {
        this.genderFilterStrLiveData.setValue(genderFilterStr);

        int filterId = switch (genderFilterStr) {
            case Gender.ANIMATE_MASCULINE -> R.id.radioAnimateMasculine;
            case Gender.INANIMATE_MASCULINE -> R.id.radioInanimateMasculine;
            case Gender.FEMININE -> R.id.radioFeminine;
            case Gender.NEUTER -> R.id.radioNeuter;
            default -> -1;
        };
        this.genderFilterIdLiveData.setValue(filterId);
    }

    public void putWordToErrorMap(String word, int errorCount) {
        if (word != null && !word.isEmpty()) {
            Map<String, Integer> newMap = new HashMap<>(getWordsWithErrors());
            newMap.put(word, errorCount);
            wordsWithErrorsLiveData.setValue(newMap);
        }
    }

    public void removeWordFromErrorMap(String word) {
        if (word != null && !word.isEmpty()) {
            Map<String, Integer> newMap = new HashMap<>(getWordsWithErrors());
            newMap.remove(word);
            wordsWithErrorsLiveData.setValue(newMap);
        }
    }
}

File: app/src/main/kotlin/com/usharik/app/AppState.kt
package com.usharik.app

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Application-wide state with Compose-ready observable state. */
class AppState {
    private val _wordsWithErrors = MutableStateFlow<Map<String, Int>>(emptyMap())
    val wordsWithErrorsFlow: StateFlow<Map<String, Int>> = _wordsWithErrors.asStateFlow()

    private val _genderFilter = MutableStateFlow(Gender.ALL)
    val genderFilterFlow: StateFlow<String> = _genderFilter.asStateFlow()

    private val _switchOffAnimation = MutableStateFlow(false)
    val switchOffAnimationFlow: StateFlow<Boolean> = _switchOffAnimation.asStateFlow()

    // Handbook selection, kept here (rather than remembered inside HandbookScreen) so it
    // survives the screen being torn down and recreated - e.g. swiping the quiz's handbook
    // overlay closed and reopening it, or navigating away from the standalone Handbook
    // destination and back - instead of resetting to the masculine "pán" default each time.
    private val _handbookGender = MutableStateFlow(DEFAULT_HANDBOOK_GENDER)
    val handbookGenderFlow: StateFlow<String> = _handbookGender.asStateFlow()
    private val _handbookParadigmByGender = MutableStateFlow(mapOf(DEFAULT_HANDBOOK_GENDER to DEFAULT_HANDBOOK_PARADIGM))
    val handbookParadigmByGenderFlow: StateFlow<Map<String, String>> = _handbookParadigmByGender.asStateFlow()

    fun getWordsWithErrors(): Map<String, Int> = _wordsWithErrors.value
    fun getGenderFilterStr(): String = _genderFilter.value
    fun getSwitchOffAnimation(): Boolean = _switchOffAnimation.value
    fun getHandbookGender(): String = _handbookGender.value
    fun getHandbookParadigmByGender(): Map<String, String> = _handbookParadigmByGender.value
    fun setWordsWithErrors(value: Map<String, Int>?) { _wordsWithErrors.value = value?.toMap().orEmpty() }
    fun setSwitchOffAnimation(value: Boolean) { _switchOffAnimation.value = value }
    fun setGenderFilterStr(value: String?) { _genderFilter.value = value ?: Gender.ALL }
    fun setHandbookGender(value: String) { _handbookGender.value = value }
    fun setHandbookParadigm(gender: String, paradigm: String) {
        _handbookParadigmByGender.value = _handbookParadigmByGender.value + (gender to paradigm)
    }
    fun putWordToErrorMap(word: String?, errorCount: Int) {
        if (!word.isNullOrBlank()) _wordsWithErrors.value = _wordsWithErrors.value + (word to errorCount)
    }
    fun removeWordFromErrorMap(word: String?) {
        if (!word.isNullOrBlank()) _wordsWithErrors.value = _wordsWithErrors.value - word
    }

    companion object {
        const val DEFAULT_HANDBOOK_GENDER = "MASCULINE"
        const val DEFAULT_HANDBOOK_PARADIGM = "pán"
    }
}
